primitives/session/src/lib.rs

```diff
@@ -106,6 +108,69 @@ impl GetValidatorCount for MembershipProof {
 	}
 }
 
+/// A trait for online node inspection in a session.
+///
+/// This trait is used to decouple the pallet-session dependency from im-online
+/// module so that the user of im-online & offences modules can pass any list of
+/// validators that are considered to be online in each session, particularly useful
+/// for the Substrate-based projects having their own staking implementation
+/// instead of using pallet-staking directly.
+pub trait ValidatorSet<AccountId> {
+	// TODO [ToDr] This could use `frame_support::Parameter` instead,although don't know if such
+	// import is legal.
+	type ValidatorId: codec::Codec + codec::EncodeLike + Clone + Eq + sp_std::fmt::Debug;
+	// TODO [ToDr] This is most likely not needed along with `AccountId`
+	type ValidatorIdOf: Convert<AccountId, Option<Self::ValidatorId>>;
+
+	/// Returns current session index.
+	fn current_index() -> SessionIndex;
+
+	/// Returns all the validators ought to be online in a session.
+	///
+	/// The returned validators are all expected to be running an authority node.
+	fn validators() -> Vec<Self::ValidatorId>;
+}
+
+/// `ValidatorSet` combined with identification type for pallet-session-historical module.
+pub trait ValidatorSetWithIdentification<AccountId>: ValidatorSet<AccountId> {
+	type Identification: codec::Codec + codec::EncodeLike + Clone + Eq + sp_std::fmt::Debug;
+	type IdentificationOf: Convert<Self::ValidatorId, Option<Self::Identification>>;
+}
+
+/// A session handler for specific key type.
+pub trait OneSessionHandler<ValidatorId>: BoundToRuntimeAppPublic {
+	/// The key type expected.
+	type Key: Decode + Default + RuntimeAppPublic;
+
+	fn on_genesis_session<'a, I: 'a>(validators: I)
```

Comment:
I know there was no docs before, but I'd appreciate if you could add some. AFAIK this hooks get's called only once int he entire lifetime of the trait - during genesis config initialisation.